File: video/ai-extensions/video-63-leetcode-5/leetcode5.swift
class Solution {
    func longestPalindrome(_ s: String) -> String {
        if s.isEmpty { return "" }
        
        let characters = Array(s)
        var start = 0, end = 0
        
        for i in 0..<characters.count {
            let len1 = expandAroundCenter(characters, left: i, right: i)
            let len2 = expandAroundCenter(characters, left: i, right: i+1)
            let len = max(len1, len2)
            if len > end - start {
                start = i - (len - 1) / 2
                end = i + len / 2
            }
        }
        
        let range = start...end
        let longestPalindrome = range.map { String(characters[$0]) }.joined()
        
        return longestPalindrome
    }
    
    private func expandAroundCenter(_ chars: [Character], left: Int, right: Int) -> Int {
        var L = left, R = right
        while L >= 0 && R < chars.count && chars[L] == chars[R] {
            L -= 1
            R += 1
        }
        return R - L - 1
    }
}

File: video/ai-extensions/video-63-leetcode-5/leetcode5.ts
function longestPalindrome(s: string): string {
  if (s.length < 1) return "";
  let start = 0,
    end = 0;

  for (let i = 0; i < s.length; i++) {
    let len1 = expandAroundCenter(s, i, i);
    let len2 = expandAroundCenter(s, i, i + 1);
    let len = Math.max(len1, len2);
    if (len > end - start) {
      start = i - ((len - 1) >> 1);
      end = i + (len >> 1);
    }
  }

  return s.substring(start, end + 1);
}

function expandAroundCenter(s: string, left: number, right: number): number {
  while (left >= 0 && right < s.length && s[left] === s[right]) {
    left--;
    right++;
  }
  return right - left - 1;
}
